### ISSUE TITLE
Promise.timeout should not use Play.current

### ISSUE THREAD

**Issue by julienrf** [playframework/playframework#4229]
See [the code](https://github.com/playframework/playframework/blob/ca664a7d108c24e08e86fb815e7104fcf326bf6e/framework/src/play/src/main/scala/play/api/libs/concurrent/Promise.scala#L48).

`Play.current` is used to retrieve the `AkkaSystem` using the runtime dependency injection API.

This is a problem because it makes it impossible to use `Promise.timeout` from a program that does not use the runtime dependency injection API.

Instead, we should provide a component that takes the `AkkaSystem` as a dependency. (Note however that the `Promise` object contains just two methods… It is probably overkill to turn it into a component)

**Comment by jroper** [comment]
Yeah, I wish actually that we had a better place for `Promise.timeout` to live, preferably in Akka or Scala.

We could make it take an implicit `ActorSystem`, that's one solution.

**Comment by gmethvin** [comment]
It also seems odd that this is under something called `Promise`. What if we just made it an extension method on the akka `Scheduler`? So we would do `actorSystem.scheduler.timeout(message, duration)`.

**Comment by julienrf** [comment]
> It also seems odd that this is under something called `Promise`.

The name `Promise` is here for historical reasons (it existed in Play before the standard Scala library had `Future` and `Promise`).

> What if we just made it an extension method on the akka `Scheduler`?

Personnally I don’t like extension methods (they are hard to document and compile error are harder to debug).

Nevertheless we can use another name than `Promise`.

**Comment by gmethvin** [comment]
Also Akka has something almost identical
```scala
import akka.pattern.after
val future = after(duration, actorSystem.scheduler)(Future(message))
```
does the same thing as
```scala
val future = Promise.timeout(message, duration)
```

**Comment by jroper** [comment]
Ok, let's deprecate `Promise.timeout` in favour of `akka.pattern.after`.

**Comment by brikis98** [comment]
It took me a while to figure out that `Promise.timeout` doesn't work if you use compile-time dependency injection because the error you get provides very little useful info:

```
java.lang.InstantiationException: null
	at sun.reflect.InstantiationExceptionConstructorAccessorImpl.newInstance(InstantiationExceptionConstructorAccessorImpl.java:48) ~[na:1.8.0_05]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:408) ~[na:1.8.0_05]
	at java.lang.Class.newInstance(Class.java:433) ~[na:1.8.0_05]
	at play.api.inject.NewInstanceInjector$.instanceOf(Injector.scala:49) ~[play_2.11-2.4.0.jar:2.4.0]
	at play.api.inject.SimpleInjector$$anonfun$instanceOf$1.apply(Injector.scala:85) ~[play_2.11-2.4.0.jar:2.4.0]
	at scala.collection.MapLike$class.getOrElse(MapLike.scala:128) ~[scala-library-2.11.6.jar:na]
	at scala.collection.AbstractMap.getOrElse(Map.scala:59) ~[scala-library-2.11.6.jar:na]
	at play.api.inject.SimpleInjector.instanceOf(Injector.scala:85) ~[play_2.11-2.4.0.jar:2.4.0]
	at play.api.inject.SimpleInjector.instanceOf(Injector.scala:80) ~[play_2.11-2.4.0.jar:2.4.0]
	at play.api.Application$$anonfun$instanceCache$1.apply(Application.scala:234) ~[play_2.11-2.4.0.jar:2.4.0]
```

Hopefully, people Googling that error in the future will come across this thread and see the suggestion to use `akka.pattern.after`.

**Comment by marcospereira** [comment]
@julienrf I think this was fixed by #4782, right? Can we close this issue?

**Comment by julienrf** [comment]
Yes!
